common/network-common/src/main/java/org/apache/spark/network/crypto/GcmTransportCipher.java

```diff
@@ -0,0 +1,356 @@
+                    0);
+            plaintextBuffer =
+                    ByteBuffer.allocate(aesGcmHkdfStreaming.getPlaintextSegmentSize());
+            ciphertextBuffer =
+                    ByteBuffer.allocate(aesGcmHkdfStreaming.getCiphertextSegmentSize());
+            decrypter = aesGcmHkdfStreaming.newStreamSegmentDecrypter();
+        }
+
+        @Override
+        public void channelRead(ChannelHandlerContext ctx, Object ciphertextMessage)
+                throws GeneralSecurityException {
+            Preconditions.checkArgument(ciphertextMessage instanceof ByteBuf,
+                    "Unrecognized message type: %s",
+                    ciphertextMessage.getClass().getName());
+            ByteBuf ciphertextNettyBuf = (ByteBuf) ciphertextMessage;
+            // The format of the output is:
+            // [8 byte length][Internal IV and header][Ciphertext][Auth Tag]
+            try {
+                while (ciphertextNettyBuf.readableBytes() > 0) {
+                    // Check if the expected ciphertext length has been read.
+                    if (expectedLength < 0 &&
+                            ciphertextNettyBuf.readableBytes() >= LENGTH_HEADER_BYTES) {
+                        expectedLength = ciphertextNettyBuf.readLong();
+                        if (expectedLength < 0) {
+                            throw new IllegalStateException("Invalid expected ciphertext length.");
+                        }
+                        ciphertextRead += LENGTH_HEADER_BYTES;
+                    }
+                    int headerLength = aesGcmHkdfStreaming.getHeaderLength();
+                    // Check if the ciphertext header has been read. This contains
+                    // the IV and other internal metadata.
+                    if (!decrypterInit &&
+                            ciphertextNettyBuf.readableBytes() >= headerLength) {
+                        ByteBuffer headerBuffer = ByteBuffer.allocate(headerLength);
+                        ciphertextNettyBuf.readBytes(headerBuffer);
+                        headerBuffer.flip();
+                        decrypter.init(headerBuffer, DEFAULT_AAD);
+                        decrypterInit = true;
+                        ciphertextRead += headerLength;
+                    }
+                    // This may occur if there weren't enough readable bytes to read the header.
+                    if (!decrypterInit) {
+                        return;
+                    }
+                    // This may occur if the expected length is just the header.
+                    if (expectedLength == ciphertextRead) {
+                        return;
+                    }
+                    ciphertextBuffer.clear();
+                    // Read the ciphertext into the local buffer
+                    int readableBytes = Integer.min(
+                            ciphertextNettyBuf.readableBytes(),
+                            ciphertextBuffer.remaining());
```

Comment:
QQ: Do we want to enforce we are reading only upto `expectedLength` ? (currently we throw an exception below if we end up reading more ...).
I am assuming it is possible for input to have more in case multiple messages are being encrypted one after another ?